### ISSUE TITLE
Cannot clean up installation after upgrade from 2.3.0 to 2.3.1

### ISSUE THREAD

**Issue by mm0ntg0m** [apache/couchdb-pkg#46]
Upgraded from CouchDB 2.3.0 to CouchDB 2.3.1 on Debian 9.9

After installation the following error message appeared

apt-get upgrade
Reading package lists... Done
Building dependency tree
Reading state information... Done
Calculating upgrade... Done
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
1 not fully installed or removed.
After this operation, 0 B of additional disk space will be used.
Do you want to continue? [Y/n]
Setting up couchdb (2.3.1~stretch) ...
dpkg: error processing package couchdb (--configure):
subprocess installed post-installation script returned error exit status 128
Errors were encountered while processing:
couchdb
E: Sub-process /usr/bin/dpkg returned an error code (1)

It appears that 2.3.1 is installed because if I check the CouchDB installation with curl localhost:5984, it returns....
{"couchdb":"Welcome","version":"2.3.1","git_sha":"c298091a4","uuid":"c89c514e09cd4f0491144e72c614af91","features":["pluggable-storage-engines","scheduler"],"vendor":{"name":"The Apache Software Foundation"}}

This error is showing up in our sandbox environment and would like to get it addressed before we attempt to update our Production environment. Lest we risk torturing our customers with unknown problems......

What suggestions do you have to clear this error?

**Comment by wohali** [comment]
I will try and reproduce this issue today.

**Comment by wohali** [comment]
Hi there,

Can you please edit `/var/lib/dpkg/info/couchdb.postinst` and add the command `set -x` on line 2, then re-run and give me the debug output? Thanks.

**Comment by mm0ntg0m** [comment]
Thank you for your response. Here is the output....

apt-get upgrade
Reading package lists... Done
Building dependency tree
Reading state information... Done
Calculating upgrade... Done
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
1 not fully installed or removed.
After this operation, 0 B of additional disk space will be used.
Do you want to continue? [Y/n]
Setting up couchdb (2.3.1~stretch) ...
+ set -e
+ . /usr/share/debconf/confmodule
+ [ !  ]
+ PERL_DL_NONLAZY=1
+ export PERL_DL_NONLAZY
+ [  ]
+ exec /usr/share/debconf/frontend /var/lib/dpkg/info/couchdb.postinst configure 2.3.0~stretch
dpkg: error processing package couchdb (--configure):
 subprocess installed post-installation script returned error exit status 128
Errors were encountered while processing:
 couchdb
E: Sub-process /usr/bin/dpkg returned an error code (1)

**Comment by wohali** [comment]
Do you have `DEBIAN_FRONTEND=noninteractive` set in your shell?

If so, can you remove it and see what happens?

If not, what happens if you add it and re-run the command?

**Comment by mm0ntg0m** [comment]
DEBIAN_FRONTEND=noninteractive was not set in .bashrc so I set it, logged out and logged back in.

I reran the the apt-get upgrade command and got the same error.

**Comment by wohali** [comment]
Don't make that the default, you won't get pop-up menus.

I'll try and reproduce this here and see what I can do for you. So far, I've been unable to do so.

It *is* safe to `apt uninstall couchdb && apt install couchdb`, by the way. Just don't `apt purge couchdb` or you'll lose the entire `/opt/couchdb` directory, **including** your `data/` and `etc/` directories.

**Comment by mm0ntg0m** [comment]
I removed DEBIAN_FRONTEND=noninteractive from .bashrc when it did not work

"uninstall" is not an option for apt so I ran apt remove couchdb && apt install couchdb.

It return the same error as before.

Thank you for all your help

**Comment by wohali** [comment]
Right, sorry. `remove` not `uninstall`. It's been a long day. :)

Sounds like something specific in your setup is causing the problem. We'll have to work through it step by step.

What do you see if you manually run `/usr/share/debconf/frontend /var/lib/dpkg/info/couchdb.postinst configure 2.3.0~stretch` as root?

**Comment by mm0ntg0m** [comment]
I ran /usr/share/debconf/frontend /var/lib/dpkg/info/couchdb.postinst configure 2.3.0~stretch as root and it returned with no message

I reran apt-get upgrade and the error is still present

**Comment by wohali** [comment]
I'm out of time today and will have to pick this up tomorrow, sorry.

**Comment by wohali** [comment]
FYI: https://github.com/apache/couchdb-fauxton/issues/1189#issuecomment-507955531

I again hope to have time to reproduce this today if possible, but we may still be in the territory of something specific on your machine being broken.

**Comment by wohali** [comment]
Hi there,

I tried to reproduce this on a fresh Debian 9.9 install today and failed. Here's the steps I took:

```
# ensure the system is up to date
apt update && apt upgrade
# now, set up the CouchDB repo per installation instructions
# install CouchDB 2.3.0, setting up as standalone
apt install couchdb=2.3.0~stretch
# then, upgrade to CouchDB 2.3.1:
apt upgrade
```

What changes, if any, have you made to your `/opt/couchdb` directory? Are there symlinks in that path anywhere?

**Comment by wohali** [comment]
For completeness, I also installed 2.3.0, setting it up as clustered with no difference in result. I also tested installing 2.3.0 with `none` configuration. Same result.

:confused: :question:

**Comment by mm0ntg0m** [comment]
Thanks for digging into this issue and spending time to understand if our problem is reproducible. 

I have attached a listing of /opt/couchdb. There is a symlink data -> /var/lib/couchdb. The same symlink exists in our 2.3.0 Prod cluster. In both the Prod and Sandbox environments we have 2 zones with 3 couchdb in each zone. The couchdb servers themselves are Debian 9.9 kvm machines running in a Proxmox environment.

It appears that couchdb 2.3.1 is installed and operating on each of the machines. The only evidence that a problem exists is the complaint from dpkg on upgrade......
- dpkg: error processing package couchdb (--configure):
- subprocess installed post-installation script returned error exit status 128

Is there a way to determine the offending components causing this upgrade error and manually removing them?

listing of /opt/couchdb/:
ls -al /opt/couchdb
total 41556
drwxr-xr-x  9 couchdb couchdb     4096 Jul  2 18:12 .
drwxr-xr-x  3 root    root        4096 Jan  8 00:15 ..
drwxr-xr-x  2 root    root        4096 Jul  2 18:12 bin
lrwxrwxrwx  1 root    root          16 Mar 11 22:27 data -> /var/lib/couchdb
-rw-r-----  1 couchdb couchdb 42411438 Jul  2 18:12 erl_crash.dump
drwxr-xr-x  8 root    root        4096 Jul  2 18:12 erts-8.3.5
drwxr-xr-x  4 couchdb couchdb     4096 Jul  2 18:12 etc
drwxr-xr-x 45 root    root        4096 Jul  2 18:12 lib
-rw-r--r--  1 root    root      100044 Mar 11 22:27 LICENSE
drwxr-xr-x  3 root    root        4096 Jul  2 18:12 releases
drwxr-xr-x  5 root    root        4096 Jul  2 18:12 share
drwxr-xr-x  2 root    root        4096 Jul  2 18:12 var

**Comment by mm0ntg0m** [comment]
It just occurred to me that when we upgraded from couchdb 2.3.0 to couchdb 2.3.1, we also upgraded from Debian 9.8 to Debian 9.9

Looking at http://ftp.debian.org/debian/dists/stretch/ChangeLog there were changes made to dpkg in the 9.9 release. 

Could this be the culprit?

**Comment by mm0ntg0m** [comment]
Update...

I cloned the CouchDB VM template that we started with to build our Sandbox and Prod clusters. The VM template was Debian 9.8 and CouchDB 2.3.0. 

I updated only couchdb 2.3.0 to couchdb 2.3.1. There were no issues.

I then upgraded all packages including the update from Debian 9.8 to Debian 9.9. There were no issues.

Going forward, I have a plan for updating our Production cluster (couchdb 2.3.1 first then everything else including Debian 9.9)

This seems like a compatibility issue between couchdb and dpkg. How do we report this problem so it can be fixed?
